File: test/rust/crates/specgate-ctsc-fixtures/src/registry/focused/rich.rs
use specgate::{SpecEvent, spec_operation};
use std::collections::{BTreeMap, BTreeSet};

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.rich")]
pub struct Address {
    #[spec_event]
    pub street: String,
    #[spec_event]
    pub city: String,
}

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.rich")]
pub struct Person {
    #[spec_event]
    pub name: String,
    #[spec_event]
    pub address: Address,
    #[spec_event]
    pub tags: Vec<String>,
    #[spec_event]
    pub scores: BTreeMap<String, i32>,
    #[spec_event]
    pub aliases: BTreeSet<String>,
    #[spec_event]
    pub nickname: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.rich")]
pub enum Shape {
    Circle { radius: i32 },
    Rectangle { width: i32, height: i32 },
    Point,
}

#[spec_operation("describe", spec = "fixture.rich")]
pub fn describe(person: Person, fallback: Option<Shape>) -> Option<String> {
    tag_count(person.tags.clone());
    fallback.map(|shape| format!("{}:{shape:?}", person.name))
}

#[spec_operation("tag_count", spec = "fixture.rich")]
pub fn tag_count(tags: Vec<String>) -> i32 {
    i32::try_from(tags.len()).unwrap_or(i32::MAX)
}

#[test]
fn rich_values_project_natively() {
    let person = Person {
        name: "Ada".to_string(),
        address: Address {
            street: "1 Main".to_string(),
            city: "London".to_string(),
        },
        tags: vec!["engineer".to_string()],
        scores: BTreeMap::from([("quality".to_string(), 10)]),
        aliases: BTreeSet::from(["A".to_string()]),
        nickname: None,
    };
    assert_eq!(tag_count(person.tags.clone()), 1);
    assert_eq!(
        describe(person, Some(Shape::Point)),
        Some("Ada:Point".to_string())
    );
}

File: test/csharp/SpecGate.CtscFixtures/Registry/Focused/Rich/Rich.cs
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Registry.Focused.Rich;

[SpecEvent("Address")]
public sealed class Address
{
    [SpecEvent("street")]
    public string Street { get; set; } = string.Empty;

    [SpecEvent("city")]
    public string City { get; set; } = string.Empty;
}

[SpecEvent("Person")]
public sealed class Person
{
    [SpecEvent("name")]
    public string Name { get; set; } = string.Empty;

    [SpecEvent("address")]
    public Address Address { get; set; } = new();

    [SpecEvent("tags")]
    public List<string> Tags { get; set; } = [];

    [SpecEvent("scores")]
    public SortedDictionary<string, int> Scores { get; set; } = [];

    [SpecEvent("aliases")]
    public SortedSet<string> Aliases { get; set; } = [];

    [SpecEvent("nickname")]
    public Option<string> Nickname { get; set; } = Option<string>.None();
}

[SpecEvent("Shape")]
public abstract class Shape;

[SpecEvent("Circle")]
public sealed class Circle : Shape
{
    [SpecEvent("radius")]
    public int Radius { get; set; }
}

[SpecEvent("Rectangle")]
public sealed class Rectangle : Shape
{
    [SpecEvent("width")]
    public int Width { get; set; }

    [SpecEvent("height")]
    public int Height { get; set; }
}

[SpecEvent("Point")]
public sealed class Point : Shape;

public static class Rich
{
    [SpecOperation("describe", Spec = "fixture.rich")]
    public static Option<string> Describe(
        [SpecInput("person")] Person person,
        [SpecInput("fallback")] Option<Shape> fallback)
    {
        _ = TagCount(person.Tags);
        return fallback.HasValue
            ? Option<string>.Some(person.Name)
            : Option<string>.None();
    }

    [SpecOperation("tag_count", Spec = "fixture.rich")]
    public static int TagCount([SpecInput("tags")] List<string> tags) => tags.Count;
}
